src/apps/inventory/mixins.py

```diff
@@ -22,3 +22,54 @@ def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
         context["most_viewed_products"] = self.get_most_viewed_products()
         return context
+
+
+class LastViewedProductsMixin:
+    def get_last_viewed_products(self):
+        redis = get_redis_connection("redis_cache")
+        user = self.request.user
+        if user.is_authenticated:
+            user_key = f"user:{user.uuid}:recentviews"
+        else:
+            user_key = f"session:{self.request.session.session_key}:recentviews"
+
+        last_viewed_products_ids = redis.zrevrange(user_key, 0, 2)
+        products = Product.objects.in_bulk(map(int, last_viewed_products_ids))
+        ordered_products = OrderedDict(
+            (int(product_id), products[int(product_id)]) for product_id in last_viewed_products_ids
```

Comment:
Why so complicated? You already extracted products and mapped them to ids with in_bulk query